feat: Add possibility to use httproute resource

Kaswob opened the pull request:
Add a possibilty to use Gateway API HTTPRoute
Add unit tests
Update schema
Update default values

Resolves https://github.com/hashicorp/vault-helm/issues/1044

## PCI review checklist

<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->

- [x] I have documented a clear reason for, and description of, the change I am making.

- [x] If applicable, I've documented a plan to revert these changes if they require more than reverting the pull request.

- [x] If applicable, I've documented the impact of any changes to security controls.

  Examples of changes to security controls include using new access control methods, adding or removing logging pipelines, etc.

Kaswob committed:
feat: Add possibility to use httproute resource

Add a possibilty to use Gateway API HTTPRoute instead of Ingress
Add unit tests
Update schema
Update default values

Kaswob committed:
feat: Add possibility to use httproute resource

Add a possibilty to use Gateway API HTTPRoute instead of Ingress
Add unit tests
Update schema
Update default values

timetofaceit commented:
@tvoran sorry for ping, but maybe you can have a look at this? Would be great to start migrating officially.

MissakaI commented:
Hi,

Sorry again for the ping. Still waiting for the official migrations.

As for the K8s docs [https://kubernetes.io/docs/concepts/services-networking/ingress/](https://kubernetes.io/docs/concepts/services-networking/ingress/):

> Note:
>
> The Kubernetes project recommends using [Gateway](https://gateway-api.sigs.k8s.io/) instead of [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/). The Ingress API has been frozen.
> 
> This means that:
> 
> - The Ingress API is generally available, and is subject to the [stability guarantees](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api) for generally available APIs. The Kubernetes project has no plans to remove Ingress from Kubernetes.
> 
> - The Ingress API is no longer being developed, and will have no further changes or updates made to it.

And official ingress-nginx

> 
>
> ## Retiring
>
> [What You Need to Know about Ingress NGINX Retirement](https://www.kubernetes.io/blog/2025/11/11/ingress-nginx-retirement/):
>
> * Best-effort maintenance will continue until March 2026.
> * Afterward, there will be no further releases, no bugfixes, and no updates to resolve any security vulnerabilities that may be discovered.
> * Existing deployments of Ingress NGINX will not be broken.
> * Existing project artifacts such as Helm charts and container images will remain available.

Capability to do the migration so from the official chart is much appreciated.

hashicorp-copywrite[bot] committed:
[COMPLIANCE] Add/Update Copyright Headers (#1184)

Co-authored-by: hashicorp-copywrite[bot] <110428419+hashicorp-copywrite[bot]@users.noreply.github.com>

vijayavelsekar committed:
Prepare release v0.33.0 (#1189)

* VAULT-44409 Default version update for k8s vault and csi provider

* VAULT-44409 version update

* VAULT-44409 Version update

* VAULT-44409 Updated versions

* VAULT-44409 Update dependency for actions/checkout

* VAULT-44409 Change log updates for dependabot pr's

* VAULT-44409 Change log minor verbiage update

* VAULT-44409 Change log updates with reference PR details

* VAULT-44409 Remove unwanted entry from change log

---------

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

johnjosephg commented on test/unit/server-httproute.bats:
```diff
@@ -0,0 +1,364 @@
+#!/usr/bin/env bats
+
```
Could we have some documentation for all of the tests written?

Kaswob replied:
Thanks for the review. Sure, docs added, tests and schema fixed. Also did make parentRefs required - it has to be specified so the HTTPRoute can attach to a Gateway.

johnjosephg commented on values.schema.json:
```diff
@@ -857,6 +857,44 @@
+                        "matches": {
+                            "type": [
+                                "array",
+                                "object"
```
matches must be an array, it cannot be both, as defined in values.yaml file. 
The corresponding changes must be made in the test file as well.

johnjosephg commented on test/unit/server-httproute.bats:
```diff
@@ -0,0 +1,364 @@
+      --show-only templates/server-httproute.yaml \
+      --set 'server.httproute.enabled=true' \
+      --set 'server.httproute.hostnames[0]=test.com' \
+      --set 'server.httproute.matches.path[0].type=PathPrefix' \
```
instead of having this as an object, where path is the array, we should have matches as the array as shown.
```
--set 'server.httproute.matches[0].path.type=PathPrefix'
```
Please make the changes wherever applicable.

Kaswob committed:
feat(httproute): make parentRefs required and improve tests

Add a mechanism which fails rendering when httproute.parentRefs is not
set - route without a parentRef caannot attach to any Gateway.
Update unit tests. Add test docs f or httproute.
Fix values schema - matches is array-only

johnjosephg reviewed (changes requested):
Could you please also add http route to an acceptance test, just to make sure that this new resource works with the existing vault setup?

Kaswob committed:
test(httproute): Add acceptance tests for HTTPRoute resource creation

Add acceptance tests that install the chart with the httproute resource.
Tests cover namespace, parentRefs, hostname, labels &
disabled-by-default case.

johnjosephg reviewed (dismissed):
LGTM

Kaswob committed:
docs: Add changelog entry

Add changelog

Kaswob committed:
Merge branch 'main' into kaswob_gateway_api
